Sources/NIOHTTP1/HTTPDecoder.swift

```diff
@@ -358,91 +376,120 @@ public class HTTPDecoder<HTTPMessageT>: ByteToMessageDecoder, AnyHTTPDecoder {
+        if self.cumulationBuffer == nil {
+            self.cumulationBuffer = buffer
+        } else {
+            self.cumulationBuffer!.write(buffer: &buffer)
+        }
 
-            let result = state.baseAddress!.withMemoryRebound(to: Int8.self, capacity: pointer.count, { p in
-                c_nio_http_parser_execute(&parser, &settings, p.advanced(by: buffer.readerIndex), buffer.readableBytes)
-            })
+        do {
+            repeat {
+                // We need to refetch the cumulationBuffer on each loop as it may has changed due re-entrance calls of channelRead(...)
+                let bufferSlice = self.cumulationBuffer!
 
-            if let error = state.currentError {
-                throw error
-            }
+                let result = try bufferSlice.withVeryUnsafeBytes { (pointer) -> size_t in
+                    state.baseAddress = pointer.baseAddress!.assumingMemoryBound(to: UInt8.self)
+
+                    let result = state.baseAddress!.withMemoryRebound(to: Int8.self, capacity: pointer.count, { p in
+                        c_nio_http_parser_execute(&parser, &settings, p.advanced(by: bufferSlice.readerIndex), bufferSlice.readableBytes)
+                    })
 
-            state.baseAddress = nil
+                    state.baseAddress = nil
 
-            let httpError = parser.http_errno
-            if httpError != 0 {
-                self.state.currentError = HTTPParserError.httpError(fromCHTTPParserErrno: http_errno(rawValue: httpError))!
-                throw self.state.currentError!
+                    if let error = state.currentError {
+                        throw error
+                    }
+
+                    let httpError = parser.http_errno
+                    if httpError != 0 {
+                        self.state.currentError = HTTPParserError.httpError(fromCHTTPParserErrno: http_errno(rawValue: httpError))!
+                        throw self.state.currentError!
+                    }
+                    return result
+                }
+
+                guard self.cumulationBuffer != nil else {
+                    // Guard against the case of closing the channel while still in the decode loop. In this case the cumulationBuffer will be nil.
+                    return
+                }
+
+                /// http_parser_execute(...) should always consume the whole buffer.
+                assert(result == bufferSlice.readableBytes)
+
+                // Update readerIndex of the cumulationBuffer itself as we will refetch it in the next loop run if needed.
+                self.cumulationBuffer!.moveReaderIndex(forwardBy: result)
+
+                if bufferSlice.readableBytes == self.cumulationBuffer!.readableBytes + result {
+                    // There was no change to the readable bytes of the cumulationBuffer by a re-entrant call of channelRead. Its safe to break the loop.
+                    break
+                }
+            } while true
```

Comment:
`repeat...while true` is a strange idiom. I know you didn't add it, but can we swap this to just `while true { }`?

Reply:
Actually, no, let's use `while var bufferSlice = self.cumulationBuffer, bufferSlice.readableBytes > 0`. That will cover your loop exit condition at the bottom and also create the buffer slice we need.